common/remote/sockfile.cpp

```diff
@@ -3585,109 +3831,141 @@ class CRemoteDiskReadActivity : public CSimpleInterfaceOf<IRemoteActivity>
-            }
-            else
-            {
-                iFileIO.setown(iFile->open(IFOread));
-                if (!iFileIO)
-                    throw MakeStringException(0, "Failed to open: '%s'", fileName);
-            }
-
-            inputStream.setown(createFileSerialStream(iFileIO, startPos));
-            prefetchBuffer.setStream(inputStream);
-            prefetcher.setown(helper->queryDiskRecordSize()->createDiskPrefetcher());
-
-            outBuilder = new MemoryBufferBuilder(resultBuffer, helper->queryOutputMeta()->getMinRecordSize());
-            chooseN = helper->getChooseNLimit();
-            limit = helper->getRowLimit();
+            iFileIO.setown(iFile->open(IFOread));
+            if (!iFileIO)
+                throw MakeStringException(1, "Failed to open: '%s'", fileName);
         }
 
+        inputStream.setown(createFileSerialStream(iFileIO, startPos));
+        prefetchBuffer.setStream(inputStream);
+        prefetcher.setown(helper->queryDiskRecordSize()->createDiskPrefetcher());
+
+        chooseN = helper->getChooseNLimit();
+        limit = helper->getRowLimit();
         opened = true;
     }
     void close()
     {
         iFileIO.clear();
         opened = false;
     }
-    bool segMonitorsMatch(const void *row) { return true; }
+    inline bool fieldFilterMatch(const void * buffer)
+    {
+        if (actualFilter.numFilterFields())
+        {
+            filterRow->setRow(buffer, 0);
+            return actualFilter.matches(*filterRow);
+        }
+        else
+            return true;
+    }
 public:
-    CRemoteDiskReadActivity(IHThorDiskReadArg &_helper, bool _compressed)
-        : compressed(_compressed), helper(&_helper), prefetchBuffer(nullptr)
+    CRemoteDiskReadActivity(IHThorDiskReadArg &_helper, bool _compressed, bool _inputGrouped, bool _outputGrouped)
+        : compressed(_compressed), inputGrouped(_inputGrouped), outputGrouped(_outputGrouped), helper(&_helper), prefetchBuffer(nullptr)
     {
         outMeta.set(helper->queryOutputMeta());
         canMatchAny = helper->canMatchAny();
+        record = &helper->queryDiskRecordSize()->queryRecordAccessor(true);
     }
     ~CRemoteDiskReadActivity()
     {
-        if (outBuilder)
-            delete outBuilder;
+        delete filterRow;
+    }
+    void addFilter(const char *filter)
+    {
+        actualFilter.addFilter(*record, filter);
+        if (!filterRow)
+            filterRow = new RtlDynRow(*record);
     }
-    virtual const void *nextRow(size32_t &retSz) override
+// IRemoteActivity impl.
+    virtual const void *nextRow(MemoryBufferBuilder &outBuilder, size32_t &retSz) override
     {
+        if (eogPending)
+        {
+            eogPending = false;
+            someInGroup = false;
+            return nullptr;
+        }
         checkOpen();
-        if (needTransform)
+        while (!eofSeen && ((chooseN == 0) || (processed < chooseN)))
```

Comment:
Are you sure the chooseN == 0 check is correct?

Reply:
Well it was code copied from hthor and matches thor behaviour too.
But it turns out 0 *should* mean 0 and not all as it is currently implemented here, in hthor and thor.
The behaviour was changed some years ago, but it looks like hthor and thor were never changed, however they get away with it, because the codegen always generates a subsequent choosen activity.
I've opened a separate JIRA to change hthor and thor choosen==0 behaviour : https://track.hpccsystems.com/browse/HPCC-19440